Source/Particles/Collision/BinaryCollision/VirtualPhotonCreation.H

```diff
@@ -0,0 +1,244 @@
+        amrex::Real sampling_factor;
+        utils::parser::getWithParser(pp_species_name, "sampling_factor", sampling_factor);
+
+        amrex::Real const alpha_pi = PhysConst::alpha / MathConst::pi;
+        amrex::Real const inv_c2 = 1._rt / (PhysConst::c * PhysConst::c);
+        amrex::Real const mass = primary.getMass();
+        amrex::Real umin = 0._rt;
+
+        int const nlevs = std::max(0, primary.finestLevel()+1);
+        for (int lev = 0; lev < nlevs; ++lev) {
+#ifdef AMREX_USE_OMP
+#pragma omp parallel if (amrex::Gpu::notInLaunchRegion())
+#endif
+            for (amrex::MFIter mfi = primary.MakeMFIter(lev); mfi.isValid(); ++mfi)
+            {
+                // Notation: _vp means virtual photon
+                // Primary particles (leptons) in the current tile
+                ParticleTileType& ptile = primary.ParticlesAt(lev, mfi);
+                const auto soa = ptile.getParticleTileData();
+
+                // Number of primary particles in the current tile
+                amrex::Long const num = ptile.numParticles();
+
+                // Vector that will contain the number of virtual photons for each primary particle
+                amrex::Gpu::DeviceVector<amrex::Long> num_vp(num, 0);
+                auto* num_vp_data = num_vp.dataPtr();
+
+                // First pass: compute the number of virtual photons for each primary particle
+                // and fill the corresponding vector
+                amrex::ParallelForRNG(num,
+                [=] AMREX_GPU_DEVICE (amrex::Long i, amrex::RandomEngine const& engine) noexcept
+                {
+                    amrex::ParticleReal ux = soa.m_rdata[PIdx::ux][i]; // u=v*gamma=p/m_e
+                    amrex::ParticleReal uy = soa.m_rdata[PIdx::uy][i];
+                    amrex::ParticleReal uz = soa.m_rdata[PIdx::uz][i];
+
+                    // Similar to Jackson's formula 15.58, but neglect longitudinal field, assume relativistic velocities, and integrate in energy
+                    amrex::ParticleReal gamma = std::sqrt( 1.0_rt +  (ux*ux + uy*uy + uz*uz) * inv_c2 );
+                    // Minimum fractional (w.r.t. the primary) photon energy
+                    amrex::ParticleReal y_min = vphoton_min_energy * inv_c2 / (gamma * mass);
+                    amrex::ParticleReal lny = std::log( y_min );
+                    // Number of virtual photons per primary particle
+                    amrex::Real r_photons = alpha_pi * lny * lny * sampling_factor;
+                    amrex::Real i_photons; // integer part
+                    amrex::Real f_photons = std::modf(r_photons, &i_photons); // fractional part
+                    amrex::Long n_photons = (amrex::Long)i_photons;
+
+                    // Account for the fractional part of r_photons by
+                    // randomly adding an extra virtual photon
+                    if(amrex::Random(engine) < f_photons){ n_photons += amrex::Long(1);}
```

Comment:
I think that this code can be simplified with 
```suggestion
                    // `n_photons` must be an integer, but must average to `r_photons` over many realizations
                    // This is achieved by adding a random number between 0 and 1, and taking the integer part.
                    amrex::Long n_photons = (amrex::Long)( r_photons + amrex::Random(engine) );
```